[Arm64] Implement Simd.Extract +

sdmaclea opened the pull request.

sdmaclea commented:
Tested with new tests added to #16008

sdmaclea commented on src/jit/codegenarm64.cpp:
```diff
@@ -5059,9 +5059,104 @@ void CodeGen::genHWIntrinsicSimdBinaryOp(GenTreeHWIntrinsic* node)
+
+    if (op2->isContainedIntOrIImmed())
+    {
+        int lane = op2->AsIntConCommon()->IconValue();
```
```src\jit\codegenarm64.cpp(5147): warning C4244: 'initializing': conversion from 'ssize_t' to 'int', possible loss of data [D:\j\workspace\x64_checked_w---76911707\bin\obj\Windows_NT.x64.Checked\src\jit\protononjit\protononjit.vcxproj]```

sdmaclea commented:
Added Simd.BitwiseSelect() support.
Added #16097 & #16102 to avoid merge conflicts

sdmaclea commented:
Simple rebase to fix merge conflict.

tannergooding commented on src/jit/codegenarm64.cpp:
```diff
@@ -5059,9 +5059,104 @@ void CodeGen::genHWIntrinsicSimdBinaryOp(GenTreeHWIntrinsic* node)
+        inst_JMP(EJ_jmp, labelBreakTarget);
+    }
+    genDefineTempLabel(labelBreakTarget);
+}
```
Why hardcode it, instead of defining a label per offset, so it can be dynamically sized, if required?

That is what I did for the x86 codegen (https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsiccodegenxarch.cpp#L655).

I do like the approach to making it reusable 😄

sdmaclea replied:
> Why hardcode it, instead of defining a label per offset, so it can be dynamically sized, if required?

Arm64 instructions are fixed size.  So it makes sense.  Single instruction per case is the expected use case.  (I would add a case instruction count parameter if needed).

Label per offset would require more complexity.  Either a separate direct branch table so the target address could be calculated, or an if else compare and branch chain.  Or ...

sdmaclea replied:
Looks like you chose the separate direct branch table.

sdmaclea replied:
The generated code will be smaller.  Since the jump table is not needed.

Certainly if a variable sized case was needed, We could add the jump table.  Right now I expect homogeneous cases of single instructions.

tannergooding replied:
> Looks like you chose the separate direct branch table

Yes, x86 instructions can potentially range in size from 1 byte to 16+ bytes (depending on prefixes, encoding, etc).

Although, within a given HWIntrinsic jump table, they will likely be fairly consistent.

CarolEidt replied:
It seems that for Arm64 this approach is best. If it were found to be necessary, it could be parameterized based on the instruction (though I don't see us needing this for anything that would require more than a single instruction per case).

sdmaclea commented:
@CarolEidt ping.  Can this be reviewed/merged?

CarolEidt commented on src/jit/codegenarm64.cpp:
```diff
@@ -5059,9 +5059,104 @@ void CodeGen::genHWIntrinsicSimdBinaryOp(GenTreeHWIntrinsic* node)
     genProduceReg(node);
 }
 
+template <typename HWIntrinsicSwitchCaseBody>
```
This needs a function header. See https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#94-function-header-comment

I would in particular describe the scenario this supports, as it may be confusing at first why we have a HW intrinsic switch.

CarolEidt commented on src/jit/codegenarm64.cpp:
```diff
@@ -5059,9 +5059,104 @@ void CodeGen::genHWIntrinsicSimdBinaryOp(GenTreeHWIntrinsic* node)
+    genDefineTempLabel(labelFirst);
+    for (int i = 0; i < swMax; ++i)
+    {
+        emitSwCase(i);
```
Since you are assuming that this generates a single instruction, you might do something like:
```
// This code assumes that emitSwCase() generates a single instruction.
unsigned prevInsCount = getEmitter()->emitInsCount;
for (int i = 0; i < swMax; ++i)
{
    emitSwCase(i);
    newInsCount = getEmitter()->emitInsCount;
    assert(newInsCount == (prevInsCount + 1));
    prevInsCount = newInsCount;
}
```
I *think* that `emitInsCount` is public.

CarolEidt commented on src/jit/codegenarm64.cpp:
```diff
@@ -5059,9 +5059,104 @@ void CodeGen::genHWIntrinsicSimdBinaryOp(GenTreeHWIntrinsic* node)
+    int lanes = emitTypeSize(simdType) / baseTypeSize;
+
+    auto emitSwCase = [&](int lane) {
+        assert(lane >= 0);
```
To me, `lane` is not very mnemonic, though others may disagree. Something like `caseImmediate` or `caseImm`?

sdmaclea replied:
@CarolEidt 
The immediate is the vector element lane for this instruction.  

`caseImm` ... feels too generic.

`lane` and `lanes` could become:
+ `element` & `elements`
+ `vectorElement` & `vectorElements`
+ `vectorIndex` & `vectorLength`

Preference?

CarolEidt replied:
Right - I hadn't really internalized that this is the specific case (not the general case as in `genHWIntrinsicSwitchTable()`). I think `element` and `elements` would be good.

CarolEidt commented on src/jit/codegenarm64.cpp:
```diff
@@ -5071,7 +5166,59 @@ void CodeGen::genHWIntrinsicSimdInsertOp(GenTreeHWIntrinsic* node)
+    bool     is16Byte = (node->gtSIMDSize > 8);
+    emitAttr attr     = is16Byte ? EA_16BYTE : EA_8BYTE;
+
+    // Arm64 has three bit select forms each use three source registers
```
nit: I would add a ';' after "forms':
```
// Arm64 has three bit select forms; each use three source registers
```

sdmaclea commented:
@CarolEidt @tannergooding 

Adding Unsigned compare zero lowering.  Unsigned compare zero tests are now all passing.
Added Simd.SetAllVector* support.  Tests are passing.
Added/fixed comments per @CarolEidt request
Added asserts per @CarolEidt request
Renamed lane to element per @CarolEidt request.

PTAL

tannergooding commented on src/jit/codegenarm64.cpp:
```diff
@@ -5023,11 +5050,39 @@ void CodeGen::genHWIntrinsicUnaryOp(GenTreeHWIntrinsic* node)
+//------------------------------------------------------------------------
+// genHWIntrinsicSimdBinaryOp:
+//
+// Produce code for a GT_HWIntrinsic node with form SimdBinaryOp.
```
Does the SIMD size matter?

sdmaclea replied:
No why?

tannergooding replied:
Was just wanting to make sure since there will be Vector64 and Vector128 types, and eventually Vector256+ (if SVE is supported).

sdmaclea replied:
All bets are off on SVE

tannergooding commented on src/jit/codegenarm64.cpp:
```diff
@@ -5059,21 +5114,334 @@ void CodeGen::genHWIntrinsicSimdBinaryOp(GenTreeHWIntrinsic* node)
+// need to generate functionally correct code when the operand is not constant
+//
+// This is required by the HW Intrinsic design to handle:
+//   debugger calls
```
It might be better to list this as: `to handle indirect calls, such as:`

tannergooding commented on src/jit/hwintrinsicArm64.cpp:
```diff
@@ -182,10 +202,36 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
+            // op1 is the first operand
+            // op2 is the second operand
+            // op3 is the third operand
+            op3 = impSIMDPopStack(simdType);
```
It might be good to have a general helper method for popping/validating the types, as was requested/done for x86.

sdmaclea replied:
I am not sure why.  Each form has different requirements.  Not sure how a helper would help.

Maybe `impScalarPopStack()`

tannergooding replied:
x86 has this method: https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsicxarch.cpp#L276

Which validates the type of a struct is a SIMD Type and the type of a scalar matches what the signature expects.

tannergooding replied:
I believe @CarolEidt was the one who requested we validate the type of the scalar values as well, rather than just calling `impPopStack().val`

sdmaclea replied:
I was assuming we relied on the C# (et al) compiler to validate this.  
Perhaps this is to check hand written IL.
Or perhaps it is just defensive programming,

In any case I think this can safely be a separate PR.

CarolEidt replied:
> Or perhaps it is just defensive programming,

Yes, that's the idea (that's true of many of the asserts and checks in the JIT, but it's surprising how often the "obvious" checks catch a problem.

> In any case I think this can safely be a separate PR.

Me too

tannergooding commented on src/jit/hwintrinsicArm64.h:
```diff
@@ -34,7 +34,8 @@ struct HWIntrinsicInfo
     enum Flags
     {
-        None
+        None          = 0,
```
Prefix with `HW_Flag_`?

sdmaclea replied:
It is inside `HWIntrinsicInfo` so it is `HWIntrinsicInfo::None`

tannergooding commented on src/jit/lsraarm64.cpp:
```diff
@@ -994,10 +995,44 @@ void LinearScan::TreeNodeInfoInitSIMD(GenTreeSIMD* simdTree, TreeNodeInfo* info)
+    GenTree*        op1     = intrinsicTree->gtOp.gtOp1;
+    GenTree*        op2     = intrinsicTree->gtOp.gtOp2;
+
+    if (op1->OperIs(GT_LIST))
```
`OperIsList()`?

sdmaclea replied:
Why?

tannergooding replied:
It is an explicit helper method for checking `OperIs(GT_LIST)`. We have a few of them and they look to be the most prevalent in the codebase (or at least the most prevalent in the code I've touched so far).

sdmaclea replied:
Why?  OperIs() is pretty standard in Arm64 lower.

CarolEidt replied:
FWIW I don't think we have a guidelines or even a lot of consistency on this. `OperIs()` is very useful for making checks for multiple oper values more concise and easier to read. For a single value, I think either is fine.

tannergooding replied:
I don't have a strong preference either, I was just mostly wondering why one over the other was used here.

tannergooding commented on src/jit/lsraarm64.cpp:
```diff
@@ -994,10 +995,44 @@ void LinearScan::TreeNodeInfoInitSIMD(GenTreeSIMD* simdTree, TreeNodeInfo* info)
+    }
+    else
+    {
+        info->srcCount += GetOperandInfo(op1);
```
Is ARM not going to have 0 operand nodes or just not yet?

sdmaclea replied:
I do not see a reason.  Unless we need to support barriers.

tannergooding replied:
> Unless we need to support barriers.
x86 added `StoreFence`, `LoadFence`, and `MemoryFence`. It also has a couple of helper methods (such as `Sse.SetZeroVector128`) which takes 0 args.

sdmaclea replied:
Separate PR if needed.

CarolEidt replied:
Sounds fine to me.

CarolEidt commented on src/jit/lowerarmarch.cpp:
```diff
@@ -515,6 +519,48 @@ void Lowering::LowerSIMD(GenTreeSIMD* simdNode)
+    auto intrinsicID   = node->gtHWIntrinsicId;
+    auto intrinsicInfo = comp->getHWIntrinsicInfo(node->gtHWIntrinsicId);
+
+    if ((intrinsicInfo.flags & HWIntrinsicInfo::LowerCmpUZero) && varTypeIsUnsigned(node->gtSIMDBaseType))
```
@sdmaclea - I'd love to see the additional comment, but I'm OK with merging now and you can add with another PR. Let me know.

sdmaclea commented:
@CarolEidt @tannergooding Pushed a final comment patch.  

Based on 340f055 test results above this could be merged once format checks pass.

sdmaclea commented:
All checks passed.  This can be merged.

tannergooding commented:
@sdmaclea, Thanks!
